lib/compat/wordpress-6.2/class-wp-theme-json-gutenberg.php

```diff
@@ -1192,181 +1897,1182 @@ protected static function get_property_value( $styles, $path, $theme_json = null
+				foreach ( $theme_json['styles']['blocks'][ $name ]['elements'] as $element => $node ) {
+					$nodes[] = array(
+						'path'     => array( 'styles', 'blocks', $name, 'elements', $element ),
+						'selector' => $selectors[ $name ]['elements'][ $element ],
+					);
+
+					// Handle any pseudo selectors for the element.
+					// TODO: Replace array_key_exists() with isset() check once WordPress drops
+					// support for PHP 5.6. See https://core.trac.wordpress.org/ticket/57067.
+					if ( array_key_exists( $element, static::VALID_ELEMENT_PSEUDO_SELECTORS ) ) {
+						foreach ( static::VALID_ELEMENT_PSEUDO_SELECTORS[ $element ] as $pseudo_selector ) {
+							if ( isset( $theme_json['styles']['blocks'][ $name ]['elements'][ $element ][ $pseudo_selector ] ) ) {
+								$nodes[] = array(
+									'path'     => array( 'styles', 'blocks', $name, 'elements', $element ),
+									'selector' => static::append_to_selector( $selectors[ $name ]['elements'][ $element ], $pseudo_selector ),
+								);
+							}
+						}
+					}
+				}
+			}
+		}
+
+		return $nodes;
+	}
+
+	/**
+	 * Gets the CSS rules for a particular block from theme.json.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param array $block_metadata Metadata about the block to get styles for.
+	 *
+	 * @return string Styles for the block.
+	 */
+	public function get_styles_for_block( $block_metadata ) {
+		$node             = _wp_array_get( $this->theme_json, $block_metadata['path'], array() );
+		$use_root_padding = isset( $this->theme_json['settings']['useRootPaddingAwareAlignments'] ) && true === $this->theme_json['settings']['useRootPaddingAwareAlignments'];
+		$selector         = $block_metadata['selector'];
+		$settings         = _wp_array_get( $this->theme_json, array( 'settings' ) );
+
+		/*
+		 * Process style declarations for block support features the current
+		 * block contains selectors for. Values for a feature with a custom
+		 * selector are filtered from the theme.json node before it is
+		 * processed as normal.
+		*/
+		$feature_declarations = array();
+
+		if ( ! empty( $block_metadata['features'] ) ) {
+			foreach ( $block_metadata['features'] as $feature_name => $feature_selector ) {
+				if ( ! empty( $node[ $feature_name ] ) ) {
+					// Create temporary node containing only the feature data
+					// to leverage existing `compute_style_properties` function.
+					$feature = array( $feature_name => $node[ $feature_name ] );
+					// Generate the feature's declarations only.
+					$new_feature_declarations = static::compute_style_properties( $feature, $settings, null, $this->theme_json );
+
+					// Merge new declarations with any that already exist for
+					// the feature selector. This may occur when multiple block
+					// support features use the same custom selector.
+					if ( isset( $feature_declarations[ $feature_selector ] ) ) {
+						foreach ( $new_feature_declarations as $new_feature_declaration ) {
```

Comment:
There's this lint error, meaning that `$new_feature_declaration` needs to be declared before using it:

![imagem](https://user-images.githubusercontent.com/583546/207862578-a283399b-4708-4832-b024-45c33ab60855.png)

@anton-vlasenko does this ring any bell? I presume Gutenberg and WordPress core have different lint rules but not sure where or how to tweak it ([core code](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme-json.php#L2097) for reference).

Reply:
Oh, actually. Isn't it a bug introduced in the core refactor?

Reply:
Going to make the change here, though core needs a separate PR to fix this, see https://github.com/WordPress/wordpress-develop/pull/3555/files#r1049610227

Reply:
Fixed at https://github.com/WordPress/gutenberg/pull/46579/commits/230ba7420d1f968ccbca4b3a8ff79ad546fdfe2b

There's some lint job for PHP that is missing in WordPress core, otherwise this should have been caught.

Reply:
@scruffian @glendaviesnz I see you worked on this area and the `get_styles_for_block` method is used in `gutenberg_add_global_styles_for_blocks` (`wp_add_global_styles_for_blocks` in core). Would you help me figure out what's the potential issue this may have caused? This is the context: [this PR](https://github.com/WordPress/wordpress-develop/pull/3555#discussion_r1049610227) in core modified how some arrays are populated and it was shipped in 6.1.1 according to the [track ticket](https://core.trac.wordpress.org/ticket/56974).

Reply:
@oandregal I don't remember having much to do with this area sorry, but looking at git blame it looks like @aaronrobertshaw has, so he may have some thoughts?

Reply:
If we're strictly talking about the experimental selectors for individual block support features, the `foreach` loop seems to be introduced at the point where declarations that have the same selectors are merged. (cc @aaronrobertshaw for fact check)

I just tested this functionality on this branch using the [search block](https://github.com/WordPress/gutenberg/blob/dff616f7a2dd47da06ea7ae43e80ba8a7851ad59/packages/block-library/src/search/block.json#L59).

I gave the color support the same `__experimentalSelector` as typography, then added some typography and color block styles to `core/search` in theme.json.

Looks like they're being merged fine still 

<img width="569" alt="Screen Shot 2022-12-16 at 10 42 41 am" src="https://user-images.githubusercontent.com/6458278/207990163-4598d171-d0d7-4571-9462-b099213f67bd.png">

Reply:
As @ramonjd noted, this section of code is related to merging declarations for features that have a selector that already has had declarations mapped to it by a different feature.

There was a [unit test](https://github.com/WordPress/gutenberg/blob/trunk/phpunit/class-wp-theme-json-test.php#L701) added that covered this functionality. It depends on a [block being registered](https://github.com/WordPress/gutenberg/blob/trunk/phpunit/bootstrap.php#L104) that set the same selector in two different features e.g. border and spacing. 

Due to the tests not running until after WordPress has started up, this test block had to be registered in the PHPUnit bootstrap to avoid cached block metadata. Now that there have been improvements around the block metadata caching we might be able to improve that setup some.